File: UUCoreData.m
//
//  UUCoreData.h
//  Useful Utilities - CoreData helpers
//
//	License:
//  You are free to use this code for whatever purposes you desire. The only requirement is that you smile everytime you use it.
//
//  Contact: @cheesemaker or jon@threejacks.com
//

#import "UUCoreData.h"

//If you want to provide your own logging mechanism, define UUDebugLog in your .pch
#ifndef UUDebugLog
#ifdef DEBUG
#define UUDebugLog(fmt, ...) NSLog(fmt, ##__VA_ARGS__)
#else
#define UUDebugLog(fmt, ...)
#endif
#endif

#pragma mark - NSError Private Extensions
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// UUCoreData

@interface NSError (UUCoreData)

- (void) uuLogDetailedErrors;

@end

#pragma mark - UUCoreData
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// UUCOreData

@interface UUCoreData ()

@property (nonatomic, strong) NSPersistentStoreCoordinator* storeCoordinator;
@property (nonatomic, strong) NSManagedObjectContext* mainThreadContext;

@end

@implementation UUCoreData

+ (void) configureWithStoreUrl:(NSURL*)storeUrl modelDefinitionBundle:(NSBundle*)bundle
{
    [[self sharedInstance] configureWithStoreUrl:storeUrl modelDefinitionBundle:bundle];
}

+ (NSError*) destroyStoreAtUrl:(NSURL*)storeUrl
{
    NSFileManager* fm = [NSFileManager defaultManager];
    
    NSError* err = nil;
    [fm removeItemAtURL:storeUrl error:&err];
    
    if (err)
    {
        UUDebugLog(@"Error trying to delete core data store: %@", err);
        return err;
    }
    else
    {
        UUCoreData* cd = [UUCoreData sharedInstance];
        cd.mainThreadContext = nil;
        cd.storeCoordinator = nil;
    }
    
    return nil;
}

+ (instancetype) sharedInstance
{
    static id theSharedObject = nil;
    static dispatch_once_t onceToken;
    
    dispatch_once (&onceToken, ^
    {
        theSharedObject = [[[self class] alloc] init];
    });
    
    return theSharedObject;
}

- (void) configureWithStoreUrl:(NSURL*)storeUrl modelDefinitionBundle:(NSBundle*)bundle
{
    if (!bundle)
    {
        bundle = [NSBundle mainBundle];
    }
    
    NSManagedObjectModel* managedObjectModel = [NSManagedObjectModel mergedModelFromBundles:@[bundle]];
    
    NSError* error;
    
    NSDictionary* options =
    @{
      NSMigratePersistentStoresAutomaticallyOption : @(YES),
      NSInferMappingModelAutomaticallyOption : @(YES)
      };
    
    self.storeCoordinator = [[NSPersistentStoreCoordinator alloc] initWithManagedObjectModel:managedObjectModel];
    if (![self.storeCoordinator addPersistentStoreWithType:NSSQLiteStoreType configuration:nil URL:storeUrl options:options error:&error])
    {
        UUDebugLog(@"Cannot Setup Core Data!");
        UUDebugLog(@"Error: %@", [error description]);
        exit(-1);
    }
    
    self.mainThreadContext = [[NSManagedObjectContext alloc] initWithConcurrencyType:NSMainQueueConcurrencyType];
    [self.mainThreadContext setPersistentStoreCoordinator:self.storeCoordinator];
    
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(otherContextDidSave:) name:NSManagedObjectContextDidSaveNotification object:nil];
}

+ (NSManagedObjectContext*) mainThreadContext
{
    return [[self sharedInstance] mainThreadContext];
}

- (void)otherContextDidSave:(NSNotification *)didSaveNotification
{
    NSManagedObjectContext* context = (NSManagedObjectContext *)didSaveNotification.object;
    
    NSManagedObjectContext* destContext = self.mainThreadContext;
    if (context != destContext && context.persistentStoreCoordinator == self.mainThreadContext.persistentStoreCoordinator)
    {
        [destContext performBlock:^
         {
             UUDebugLog(@"Merging changes from background context");
             [destContext mergeChangesFromContextDidSaveNotification:didSaveNotification];
         }];
    }
}

+ (NSManagedObjectContext*) workerThreadContext
{
    NSPersistentStoreCoordinator* coordinator = [[self sharedInstance] storeCoordinator];
    NSManagedObjectContext* context = [[NSManagedObjectContext alloc] initWithConcurrencyType:NSPrivateQueueConcurrencyType];
    [context setPersistentStoreCoordinator:coordinator];
    return context;
}

@end

#pragma mark - NSManagedObjectContext Extensions
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// NSManagedObjectContext extensions

@implementation NSManagedObjectContext (UUCoreData)

- (BOOL) uuSubmitChanges
{
    __block BOOL result = NO;
    
    if ([self hasChanges])
    {
        [self performBlockAndWait:^
        {
            NSError* error = nil;
            result = [self save:&error];
             
            if (!result)
            {
                [error uuLogDetailedErrors];
            }
        }];
    }
    
    return result;
}

+ (NSArray*) uuConvertObjects:(NSArray*)objects toContext:(NSManagedObjectContext*)context
{
    NSMutableArray* convertedObjects = nil;
    
    if (objects)
    {
        convertedObjects = [NSMutableArray arrayWithCapacity:objects.count];
        
        for (NSManagedObject* obj in objects)
        {
            NSManagedObject* converted = [self uuConvertObject:obj toContext:context];
            [convertedObjects addObject:converted];
        }
    }
    
    return [convertedObjects copy];
}

+ (id) uuConvertObject:(id)object toContext:(NSManagedObjectContext*)context
{
    id converted = object;
    
    if (object)
    {
        if ([object isKindOfClass:[NSManagedObject class]])
        {
            NSManagedObject* managedObj = (NSManagedObject*)object;
            converted = [context objectWithID:managedObj.objectID];
        }
        else if ([object isKindOfClass:[NSArray class]])
        {
            NSMutableArray* convertedObjects = [NSMutableArray array];
            
            NSArray* arrayOfObjects = (NSArray*)object;
            
            for (id obj in arrayOfObjects)
            {
                if ([obj isKindOfClass:[NSManagedObject class]])
                {
                    NSManagedObject* managedObj = (NSManagedObject*)obj;
                    NSManagedObject* converted = [context objectWithID:managedObj.objectID];
                    if (converted)
                    {
                        [convertedObjects addObject:converted];
                    }
                }
            }
            
            converted = [convertedObjects copy];
        }
    }
    
    return converted;
}

- (void) uuDeleteObjects:(NSArray*)list
{
    [self performBlockAndWait:^
     {
         for (NSManagedObject* obj in list)
         {
             [self deleteObject:obj];
         }
     }];
}

- (void) uuDeleteAllObjects:(void(^)(void))completion
{
    NSArray* entityList = [self.persistentStoreCoordinator.managedObjectModel.entitiesByName allKeys];
    
    [self performBlock:^
    {
        for (NSString* entityName in entityList)
        {
            NSFetchRequest* fr = [[NSFetchRequest alloc] init];
            fr. entity = [NSEntityDescription entityForName:entityName inManagedObjectContext:self];
            fr.predicate = nil;
            
            NSError* err = nil;
            NSArray* objects = [self executeFetchRequest:fr error:&err];
            for (NSManagedObject* obj in objects)
            {
                [self deleteObject:obj];
            }
        }
        
        [self uuSubmitChanges];
        
        if (completion)
        {
            dispatch_async(dispatch_get_main_queue(), completion);
        }
    }];
}

@end

#pragma mark - NSManagedObject Extensions
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// NSManagedObject extensions

@implementation NSManagedObject (UUCoreData)

+ (NSString*) uuEntityName
{
    return NSStringFromClass([self class]);
}

+ (NSFetchRequest*) uuFetchRequestInContext:(NSManagedObjectContext*)context
                                  predicate:(NSPredicate*)predicate
                            sortDescriptors:(NSArray*)sortDescriptors
                                      limit:(NSNumber*)limit
{
    return [self uuFetchRequestInContext:context predicate:predicate sortDescriptors:sortDescriptors offset:nil limit:limit];
}

+ (NSFetchRequest*) uuFetchRequestInContext:(NSManagedObjectContext*)context
                                  predicate:(NSPredicate*)predicate
                            sortDescriptors:(NSArray*)sortDescriptors
                                     offset:(NSNumber*)offset
                                      limit:(NSNumber*)limit
{
    NSFetchRequest* fetchRequest = [[NSFetchRequest alloc] init];
    fetchRequest.entity = [NSEntityDescription entityForName:[self uuEntityName] inManagedObjectContext:context];
    fetchRequest.sortDescriptors = sortDescriptors;
    fetchRequest.predicate = predicate;
    
    if (offset != nil)
    {
        [fetchRequest setFetchOffset:[offset unsignedIntegerValue]];
    }
    
    if (limit != nil)
    {
        [fetchRequest setFetchLimit:[limit unsignedIntegerValue]];
    }
    
    return fetchRequest;
}

+ (NSArray*) uuFetchObjectsWithPredicate:(NSPredicate*)predicate
                                 context:(NSManagedObjectContext*)context
{
    return [self uuFetchObjectsWithPredicate:predicate sortDescriptors:nil limit:nil context:context];
}

+ (NSArray*) uuFetchObjectsWithPredicate:(NSPredicate*)predicate
                         sortDescriptors:(NSArray*)sortDescriptors
                                   limit:(NSNumber*)limit
                                 context:(NSManagedObjectContext*)context
{
    return [self uuFetchObjectsWithPredicate:predicate sortDescriptors:sortDescriptors offset:nil limit:limit context:context];
}

+ (NSArray*) uuFetchObjectsWithPredicate:(NSPredicate*)predicate
                         sortDescriptors:(NSArray*)sortDescriptors
                                  offset:(NSNumber*)offset
                                   limit:(NSNumber*)limit
                                 context:(NSManagedObjectContext*)context
{
    NSFetchRequest* fetchRequest = [self uuFetchRequestInContext:context predicate:predicate sortDescriptors:sortDescriptors offset:offset limit:limit];
    
    __block NSArray* fetchResults = nil;
    
    [context performBlockAndWait:^
     {
         NSError* err = nil;
         fetchResults = [context executeFetchRequest:fetchRequest error:&err];
         if (err)
         {
             [err uuLogDetailedErrors];
         }
     }];
    
    
    return fetchResults;
}

+ (NSInteger) uuBatchUpdate:(NSDictionary*)properties predicate:(NSPredicate*)predicate context:(NSManagedObjectContext*)context
{
    NSBatchUpdateRequest* req = [NSBatchUpdateRequest batchUpdateRequestWithEntityName:[self uuEntityName]];
    req.predicate = predicate;
    req.propertiesToUpdate = properties;
    req.resultType = NSUpdatedObjectsCountResultType;
    
    __block NSNumber* updatedRows = 0;

    [context performBlockAndWait:^
    {
        NSError* err = nil;
        NSPersistentStoreResult* result = [context executeRequest:req error:&err];
        if (err)
        {
            [err uuLogDetailedErrors];
        }
        
        if ([result isKindOfClass:[NSBatchUpdateResult class]])
        {
            NSBatchUpdateResult* batchResult = (NSBatchUpdateResult*)result;
            updatedRows = batchResult.result;
        }
    }];
    
    return updatedRows.integerValue;
}

+ (instancetype) uuFetchSingleObjectWithPredicate:(NSPredicate*)predicate
                                          context:(NSManagedObjectContext*)context
{
    return [self uuFetchSingleObjectWithPredicate:predicate sortDescriptors:nil context:context];
}

+ (instancetype) uuFetchSingleObjectWithPredicate:(NSPredicate*)predicate
                                  sortDescriptors:(NSArray*)sortDescriptors
                                          context:(NSManagedObjectContext*)context
{
    NSArray* list = [self uuFetchObjectsWithPredicate:predicate sortDescriptors:sortDescriptors limit:@(1) context:context];
    if (list && list.count > 0)
    {
        return list[0];
    }
    else
    {
        return nil;
    }
}

+ (instancetype) uuFetchOrCreateSingleEntityWithPredicate:(NSPredicate *)predicate context:(NSManagedObjectContext *)context
{
    NSManagedObject* obj = [self uuFetchSingleObjectWithPredicate:predicate context:context];
    
    if (!obj)
    {
        obj = [NSEntityDescription insertNewObjectForEntityForName:[self uuEntityName] inManagedObjectContext:context];
    }
    
    return obj;
}

+ (void) uuDeleteObjectsWithPredicate:(NSPredicate*)predicate
                              context:(NSManagedObjectContext*)context
{
    NSArray* list = [self uuFetchObjectsWithPredicate:predicate context:context];
    for (NSManagedObject* obj in list)
    {
        [context deleteObject:obj];
    }
}

+ (NSUInteger) uuCountObjectsWithPredicate:(NSPredicate*)predicate
                                   context:(NSManagedObjectContext*)context
{
    NSFetchRequest* fetchRequest = [self uuFetchRequestInContext:context predicate:predicate sortDescriptors:nil limit:nil];
    __block NSUInteger count = 0;
    
    [context performBlockAndWait:^
    {
        NSError* err = nil;
        count = [context countForFetchRequest:fetchRequest error:&err];
        if (err)
        {
            [err uuLogDetailedErrors];
        }
    }];
    
    return count;
}

@end



#pragma mark - NSError Extensions
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// NSError extensions

@implementation NSError (UUCoreData)

- (void) uuLogDetailedErrors
{
    NSError* error = self;
    
    UUDebugLog(@"ERROR: %@", [error localizedDescription]);
    NSArray* detailedErrors = [[error userInfo] objectForKey:NSDetailedErrorsKey];
    if (detailedErrors != nil && [detailedErrors count] > 0)
    {
        for (NSError* de in detailedErrors)
        {
            UUDebugLog(@"  DetailedError: %@", [de userInfo]);
        }
    }
    else
    {
        UUDebugLog(@"  %@", [error userInfo]);
    }
}

@end

File: swift/UUToolbox/UUToolbox/UUCoreData.swift
//  UUCoreData
//  Useful Utilities - Helpful methods for Core Data
//
//	License:
//  You are free to use this code for whatever purposes you desire.
//  The only requirement is that you smile everytime you use it.
//

import UIKit
import CoreData

open class UUCoreData: NSObject
{
    public var mainThreadContext : NSManagedObjectContext?
    public var storeCoordinator : NSPersistentStoreCoordinator?
    
    public override init()
    {
        super.init()
    }
    
    public init(url: URL, modelDefinitionBundle: Bundle = Bundle.main)
    {
        super.init()
        configure(url: url, modelDefinitionBundle: modelDefinitionBundle)
    }
    
    public init(url: URL, model: NSManagedObjectModel)
    {
        super.init()
        configure(url: url, model: model)
    }
    
    private func configure(url: URL, modelDefinitionBundle: Bundle = Bundle.main)
    {
        let mom : NSManagedObjectModel? = NSManagedObjectModel.mergedModel(from: [modelDefinitionBundle])
        if (mom == nil)
        {
            UUDebugLog("WARNING! Unable to create managed object model!")
            return
        }
        
        configure(url: url, model: mom!)
    }
    
    private func configure(url: URL, model: NSManagedObjectModel)
    {
        let options : [AnyHashable:Any] =
        [
            NSMigratePersistentStoresAutomaticallyOption : true,
            NSInferredMappingModelError: true
        ]
        
        storeCoordinator = NSPersistentStoreCoordinator(managedObjectModel: model)
        
        do
        {
            try storeCoordinator?.addPersistentStore(ofType: NSSQLiteStoreType, configurationName: nil, at: url, options: options)
        }
        catch (let err)
        {
            UUDebugLog("Error setting up CoreData: %@", String(describing: err))
            
            do
            {
                try FileManager.default.removeItem(at: url)
                
                try storeCoordinator?.addPersistentStore(ofType: NSSQLiteStoreType, configurationName: nil, at: url, options: options)
            }
            catch (let innerError)
            {
                UUDebugLog("Error Clearing CoreData: %@", String(describing: innerError))
            }
        }
        
        mainThreadContext = NSManagedObjectContext(concurrencyType: .mainQueueConcurrencyType)
        mainThreadContext?.persistentStoreCoordinator = storeCoordinator
        
        NotificationCenter.default.addObserver(self, selector: #selector(otherContextDidSave), name: NSNotification.Name.NSManagedObjectContextDidSave, object: nil)
    }
    
    func shutdown()
    {
        mainThreadContext = nil
        storeCoordinator = nil
    }
    
    @objc public func otherContextDidSave(notification: Notification)
    {
        let destContext : NSManagedObjectContext? = mainThreadContext
        
        if (destContext != nil)
        {
            let savedContext : NSManagedObjectContext? = notification.object as? NSManagedObjectContext
            if (savedContext != nil)
            {
                if (savedContext != destContext && savedContext!.persistentStoreCoordinator == mainThreadContext!.persistentStoreCoordinator)
                {
                    destContext!.perform
                    {
                        UUDebugLog("Merging changes from background context")
                        destContext!.mergeChanges(fromContextDidSave: notification)
                    }
                }
            }
        }
    }
    
    public func workerThreadContext() -> NSManagedObjectContext
    {
        let moc : NSManagedObjectContext = NSManagedObjectContext(concurrencyType: .privateQueueConcurrencyType)
        moc.persistentStoreCoordinator = storeCoordinator
        return moc
    }
}

public extension NSManagedObjectContext
{
    public func uuSubmitChanges() -> Error?
    {
        var error: Error? = nil
        
        if (hasChanges)
        {
            performAndWait
            {
                do
                {
                    try self.save()
                }
                catch (let err)
                {
                    error = err
                    UUDebugLog("Error saving core data: %@", String(describing: err))
                }
            }
        }
        
        return error
    }
    
    public func uuDeleteObjects(_ list: [Any])
    {
        performAndWait
        {
            for obj in list
            {
                if (obj is NSManagedObject)
                {
                    self.delete(obj as! NSManagedObject)
                }
            }
        }
    }
    
    public func uuDeleteAllObjects()
    {
        performAndWait
        {
            let entityList = self.persistentStoreCoordinator?.managedObjectModel.entitiesByName
            
            for entity in entityList!
            {
                let fr = NSFetchRequest<NSFetchRequestResult>()
                fr.entity = NSEntityDescription.entity(forEntityName: entity.key, in: self)
                
                do
                {
                    let objects = try self.fetch(fr)
                    for obj in objects
                    {
                        if (obj is NSManagedObject)
                        {
                            self.delete(obj as! NSManagedObject)
                        }
                    }
                }
                catch (let err)
                {
                    UUDebugLog("Error deleting all objects: %@", String(describing: err))
                }
            }
        }
    }
}

extension NSError
{
    func uuLogDetailedErrors()
    {
        UUDebugLog("ERROR: %@", localizedDescription)
        
        let detailedErrors = userInfo[NSDetailedErrorsKey] as? [NSError]
        if (detailedErrors != nil)
        {
            for de in detailedErrors!
            {
                UUDebugLog("  DetailedError: %@", de.userInfo)
            }
        }
        else
        {
            UUDebugLog("  %@", userInfo)
        }
    }
}

public extension NSManagedObject
{
    public static var uuEntityName : String
    {
        return String(describing: self)
    }
    
    public static func uuFetchRequest(
        predicate: NSPredicate? = nil,
        sortDescriptors: [NSSortDescriptor]? = nil,
        offset: Int? = nil,
        limit: Int? = nil,
        context: NSManagedObjectContext) -> NSFetchRequest<NSFetchRequestResult>
    {
        let fr = NSFetchRequest<NSFetchRequestResult>()
        fr.entity = NSEntityDescription.entity(forEntityName: uuEntityName, in: context)
        fr.sortDescriptors = sortDescriptors
        fr.predicate = predicate
        
        if (offset != nil)
        {
            fr.fetchOffset = offset!
        }
        
        if (limit != nil)
        {
            fr.fetchLimit = limit!
        }
        
        return fr
    }
    
    public static func uuFetchObjects(
        predicate: NSPredicate? = nil,
        sortDescriptors: [NSSortDescriptor]? = nil,
        offset: Int? = nil,
        limit: Int? = nil,
        context: NSManagedObjectContext) -> [Any]
    {
        let fr = uuFetchRequest(predicate: predicate, sortDescriptors: sortDescriptors, offset: offset, limit: limit, context: context)
        return uuExecuteFetch(fetchRequest: fr, context: context)
    }
    
    public static func uuFetchDictionaries(
        predicate: NSPredicate? = nil,
        sortDescriptors: [NSSortDescriptor]? = nil,
        propertiesToFetch: [Any]? = nil,
        offset: Int? = nil,
        limit: Int? = nil,
        context: NSManagedObjectContext) -> [[AnyHashable:Any]]
    {
        let fr = uuFetchRequest(predicate: predicate, sortDescriptors: sortDescriptors, offset: offset, limit: limit, context: context)
        fr.resultType = .dictionaryResultType
        fr.propertiesToFetch = propertiesToFetch
        
        guard let result = uuExecuteFetch(fetchRequest: fr, context: context) as? [[AnyHashable:Any]] else
        {
            return []
        }
        
        return result
    }
    
    public static func uuExecuteFetch(
        fetchRequest: NSFetchRequest<NSFetchRequestResult>,
        context: NSManagedObjectContext) -> [Any]
    {
        var results : [Any] = []
        
        context.performAndWait
        {
            do
            {
                results = try context.fetch(fetchRequest)
            }
            catch (let err)
            {
                (err as NSError).uuLogDetailedErrors()
                results = []
            }
        }
        
        return results
    }
    
    public class func uuFetchSingleObject(
        predicate: NSPredicate? = nil,
        sortDescriptors: [NSSortDescriptor]? = nil,
        context: NSManagedObjectContext) -> Self?
    {
        return uuFetchSingleObjectInternal(predicate: predicate,  sortDescriptors: sortDescriptors, context: context)
    }
    
    public class func uuFetchSingleDictionary(
        predicate: NSPredicate? = nil,
        sortDescriptors: [NSSortDescriptor]? = nil,
        propertiesToFetch: [Any]? = nil,
        context: NSManagedObjectContext) -> [AnyHashable:Any]?
    {
        let list = uuFetchDictionaries(predicate: predicate, sortDescriptors: sortDescriptors, propertiesToFetch: propertiesToFetch, offset: nil, limit: 1, context: context)
        
        var single : [AnyHashable:Any]? = nil
        
        if (list.count > 0)
        {
            single = list[0]
        }
        
        return single
    }
    
    private class func uuFetchSingleObjectInternal<T>(
        predicate: NSPredicate? = nil,
        sortDescriptors: [NSSortDescriptor]? = nil,
        context: NSManagedObjectContext) -> T?
    {
        let list = uuFetchObjects(predicate: predicate, sortDescriptors: sortDescriptors, offset: nil, limit: 1, context: context)
        
        var single : Any? = nil
        
        if (list.count > 0)
        {
            single = list[0]
        }
        
        return single as? T
    }
    
    public static func uuFetchOrCreate(
        predicate: NSPredicate? = nil,
        sortDescriptors: [NSSortDescriptor]? = nil,
        context: NSManagedObjectContext) -> Self
    {
        var obj = uuFetchSingleObject(predicate: predicate, sortDescriptors: sortDescriptors, context: context)
        
        context.performAndWait
        {
            if (obj == nil)
            {
                obj = uuCreate(context: context)
            }
        }
        
        return obj!
    }
    
    public class func uuCreate(
        context: NSManagedObjectContext) -> Self
    {
        return uuCreateInternal(context: context)
    }
    
    private class func uuCreateInternal<T>(context: NSManagedObjectContext) -> T
    {
        var obj : T? = nil
        
        context.performAndWait
        {
            obj = NSEntityDescription.insertNewObject(forEntityName: uuEntityName, into: context) as? T
        }
        
        return obj!
        
    }
    
    public static func uuDeleteObjects(
        predicate: NSPredicate? = nil,
        context: NSManagedObjectContext)
    {
        context.performAndWait
        {
            let list = uuFetchObjects(predicate: predicate, context: context)
            
            for obj in list
            {
                if (obj is NSManagedObject)
                {
                    context.delete(obj as! NSManagedObject)
                }
            }
        }
    }
    
    public static func uuCountObjects(
        predicate: NSPredicate? = nil,
        context: NSManagedObjectContext) -> Int
    {
        let fr = uuFetchRequest(predicate: predicate, context: context)
        
        var count : Int = 0
        
        context.performAndWait
        {
            do
            {
                count = try context.count(for: fr)
            }
            catch (let err)
            {
                (err as NSError).uuLogDetailedErrors()
            }
        }
        
        return count
    }
    
    public static func uuLogTable(
        predicate: NSPredicate? = nil,
        sortDescriptors: [NSSortDescriptor]? = nil,
        offset: Int? = nil,
        limit: Int? = nil,
        context: NSManagedObjectContext,
        logMessage: String = "")
    {
#if DEBUG
        UUDebugLog("Log Table -- \(uuEntityName) -- \(logMessage)")
        
        let list = uuFetchObjects(predicate: predicate, sortDescriptors: sortDescriptors, offset: offset, limit: limit, context: context)
        
        UUDebugLog("There are \(list.count) records in \(uuEntityName) table")
        
        var i = 0
        for o in list
        {
            if let odbg = o as? CustomDebugStringConvertible
            {
                UUDebugLog("\(uuEntityName)-\(i): \(odbg.debugDescription)")
            }
            else
            {
                UUDebugLog("\(uuEntityName)-\(i): \(o)")
            }
            
            i = i + 1
        }
#endif
    }
}


public extension UUCoreData
{
    private static var sharedStore : UUCoreData? = nil
    
    public static func configure(url: URL, modelDefinitionBundle: Bundle = Bundle.main)
    {
        sharedStore = UUCoreData(url: url, modelDefinitionBundle: modelDefinitionBundle)
    }
    
    public static var shared : UUCoreData?
    {
        return sharedStore
    }
    
    public static var mainThreadContext: NSManagedObjectContext?
    {
        return shared?.mainThreadContext
    }
    
    public static func workerThreadContext() -> NSManagedObjectContext?
    {
        return shared?.workerThreadContext()
    }
    
    public static func destroyStore(at url: URL)
    {
        do
        {
            try FileManager.default.removeItem(at: url)
        }
        catch (let err)
        {
            UUDebugLog("Error deleting store file: \(err)")
        }
        
        shared?.shutdown()
    }
}

